KerasClassifier get_activations makes assumption about input shape

**Describe the bug**
Keras get_activations makes hard-coded assumptions about the shape of the input:

```python
assert len(x_preprocessed.shape) == 4
```

**To Reproduce**
Train a Keras classifier on MNIST, 
```python
model = Model()
# .... make a simple MNIST model and fit it
kc = KerasClassifier(model)
kc.get_activations(test_data[:1], 1)
```

That results in the following exception:
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-78-de813ce063fa> in <module>()
----> 1 kc.get_activations(test_data[:1], 1)

~/nemesis/art/classifiers/keras.py in get_activations(self, x, layer, batch_size)
    397         x_preprocessed, _ = self._apply_preprocessing(x=x_expanded, y=None, fit=False)
    398 
--> 399         assert len(x_preprocessed.shape) == 4
    400 
    401         # Determine shape of expected output and prepare array

AssertionError: 
```

**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**System information (please complete the following information):**
 - OS: Linux
 - Python version: 3.6
 - 1.0.0
 - Keras

--- a/art/attacks/__init__.py
+++ b/art/attacks/__init__.py
@@ -1,7 +1,7 @@
 """
 Module providing adversarial attacks under a common interface.
 """
-from art.attacks.attack import EvasionAttack, PoisoningAttack, ExtractionAttack
+from art.attacks.attack import Attack, EvasionAttack, PoisoningAttack, ExtractionAttack
 
 from art.attacks.evasion.adversarial_patch import AdversarialPatch
 from art.attacks.evasion.boundary import BoundaryAttack
@@ -24,3 +24,4 @@ from art.attacks.evasion.zoo import ZooAttack
 from art.attacks.poisoning.poisoning_attack_svm import PoisoningAttackSVM
 
 from art.attacks.extraction.functionally_equivalent_extraction import FunctionallyEquivalentExtraction
+from art.attacks.extraction.copycat_cnn import CopycatCNN

--- a/art/attacks/evasion/__init__.py
+++ b/art/attacks/evasion/__init__.py
@@ -1,0 +1,20 @@
+"""
+Module providing evasion attacks under a common interface.
+"""
+from art.attacks.evasion.adversarial_patch import AdversarialPatch
+from art.attacks.evasion.boundary import BoundaryAttack
+from art.attacks.evasion.carlini import CarliniL2Method, CarliniLInfMethod
+from art.attacks.evasion.decision_tree_attack import DecisionTreeAttack
+from art.attacks.evasion.deepfool import DeepFool
+from art.attacks.evasion.elastic_net import ElasticNet
+from art.attacks.evasion.fast_gradient import FastGradientMethod
+from art.attacks.evasion.hclu import HighConfidenceLowUncertainty
+from art.attacks.evasion.hop_skip_jump import HopSkipJump
+from art.attacks.evasion.iterative_method import BasicIterativeMethod
+from art.attacks.evasion.newtonfool import NewtonFool
+from art.attacks.evasion.projected_gradient_descent import ProjectedGradientDescent
+from art.attacks.evasion.saliency_map import SaliencyMapMethod
+from art.attacks.evasion.spatial_transformation import SpatialTransformation
+from art.attacks.evasion.universal_perturbation import UniversalPerturbation
+from art.attacks.evasion.virtual_adversarial import VirtualAdversarialMethod
+from art.attacks.evasion.zoo import ZooAttack

--- a/art/attacks/extraction/__init__.py
+++ b/art/attacks/extraction/__init__.py
@@ -1,0 +1,5 @@
+"""
+Module providing extraction attacks under a common interface.
+"""
+from art.attacks.extraction.functionally_equivalent_extraction import FunctionallyEquivalentExtraction
+from art.attacks.extraction.copycat_cnn import CopycatCNN

--- a/art/attacks/poisoning/__init__.py
+++ b/art/attacks/poisoning/__init__.py
@@ -1,0 +1,4 @@
+"""
+Module providing poisoning attacks under a common interface.
+"""
+from art.attacks.poisoning.poisoning_attack_svm import PoisoningAttackSVM

--- a/art/classifiers/keras.py
+++ b/art/classifiers/keras.py
@@ -432,8 +432,6 @@ class KerasClassifier(ClassifierNeuralNetwork, ClassifierGradients, Classifier):
         # Apply preprocessing
         x_preprocessed, _ = self._apply_preprocessing(x=x_expanded, y=None, fit=False)
 
-        assert len(x_preprocessed.shape) == 4
-
         # Determine shape of expected output and prepare array
         output_shape = output_func([x_preprocessed[0][None, ...]])[0].shape
         activations = np.zeros((x_preprocessed.shape[0],) + output_shape[1:], dtype=ART_NUMPY_DTYPE)

--- a/docs/index.rst
+++ b/docs/index.rst
@@ -104,6 +104,9 @@ Implemented Attacks, Defences, Detections, Metrics, Certifications and Verificat
    :caption: Modules
 
    modules/attacks
+   modules/attacks/evasion
+   modules/attacks/extraction
+   modules/attacks/poisoning
    modules/classifiers
    modules/classifiers/classifiers_scikitlearn
    modules/data_generators

--- a/docs/modules/attacks.rst
+++ b/docs/modules/attacks.rst
@@ -2,121 +2,22 @@
 ==================
 .. automodule:: art.attacks
 
-Adversarial Patch
------------------
-.. autoclass:: AdversarialPatch
-   :members:
-   :special-members:
-
-Decision-Based Attack
----------------------
-.. autoclass:: BoundaryAttack
-   :members:
-   :special-members:
-
-Carlini and Wagner L_2 Attack
------------------------------
-.. autoclass:: CarliniL2Method
-   :members:
-   :special-members:
-
-Carlini and Wagner L_inf Attack
--------------------------------
-.. autoclass:: CarliniLInfMethod
-   :members:
-   :special-members:
-
-DeepFool
---------
-.. autoclass:: DeepFool
-   :members:
-   :special-members:
-
-Elastic Net Attack (EAD)
-------------------------
-.. autoclass:: ElasticNet
-   :members:
-   :special-members:
-
-Fast Gradient Method (FGM)
---------------------------
-.. autoclass:: FastGradientMethod
-   :members:
-   :special-members:
-
-Basic Iterative Method (BIM)
-----------------------------
-.. autoclass:: BasicIterativeMethod
-   :members:
-   :special-members:
-
-Projected Gradient Descent (PGD)
---------------------------------
-.. autoclass:: ProjectedGradientDescent
-   :members:
-   :special-members:
-
-Jacobian Saliency Map Attack (JSMA)
------------------------------------
-.. autoclass:: SaliencyMapMethod
-   :members:
-   :special-members:
-
-NewtonFool
-----------
-.. autoclass:: NewtonFool
-   :members:
-   :special-members:
-
-Spatial Transformations Attack
-------------------------------
-.. autoclass:: SpatialTransformation
-   :members:
-   :special-members:
-
-SVM Attack
-----------
-.. autoclass:: SVMAttack
-   :members:
-   :special-members:
-
-Universal Perturbation Attack
------------------------------
-.. autoclass:: UniversalPerturbation
-   :members:
-   :special-members:
-
-Virtual Adversarial Method
---------------------------
-.. autoclass:: VirtualAdversarialMethod
-   :members:
-   :special-members:
-
-Zeroth-Order Optimization Attack (ZOO)
---------------------------------------
-.. autoclass:: ZooAttack
-   :members:
-   :special-members:
-
-HopSkipJump Attack
+Base Class Attacks
 ------------------
-.. autoclass:: HopSkipJump
-   :members:
-   :special-members:
-
-Decision Tree Attack
---------------------
-.. autoclass:: DecisionTreeAttack
-   :members:
-   :special-members:
-
-High Confidence Low Uncertainty Attack
---------------------------------------
-.. autoclass:: HighConfidenceLowUncertainty
-   :members:
-   :special-members:
-
-Base Class
-----------
 .. autoclass:: Attack
+   :members:
+
+Base Class Evasion Attacks
+--------------------------
+.. autoclass:: EvasionAttack
+   :members:
+
+Base Class Poisoning Attacks
+----------------------------
+.. autoclass:: PoisoningAttack
+   :members:
+
+Base Class Extraction Attacks
+-----------------------------
+.. autoclass:: ExtractionAttack
    :members:

--- a/docs/modules/attacks/evasion.rst
+++ b/docs/modules/attacks/evasion.rst
@@ -1,0 +1,111 @@
+:mod:`art.attacks.evasion`
+==========================
+.. automodule:: art.attacks.evasion
+
+Adversarial Patch
+-----------------
+.. autoclass:: AdversarialPatch
+   :members:
+   :special-members:
+
+Decision-Based Attack
+---------------------
+.. autoclass:: BoundaryAttack
+   :members:
+   :special-members:
+
+Carlini and Wagner L_2 Attack
+-----------------------------
+.. autoclass:: CarliniL2Method
+   :members:
+   :special-members:
+
+Carlini and Wagner L_inf Attack
+-------------------------------
+.. autoclass:: CarliniLInfMethod
+   :members:
+   :special-members:
+
+DeepFool
+--------
+.. autoclass:: DeepFool
+   :members:
+   :special-members:
+
+Elastic Net Attack (EAD)
+------------------------
+.. autoclass:: ElasticNet
+   :members:
+   :special-members:
+
+Fast Gradient Method (FGM)
+--------------------------
+.. autoclass:: FastGradientMethod
+   :members:
+   :special-members:
+
+Basic Iterative Method (BIM)
+----------------------------
+.. autoclass:: BasicIterativeMethod
+   :members:
+   :special-members:
+
+Projected Gradient Descent (PGD)
+--------------------------------
+.. autoclass:: ProjectedGradientDescent
+   :members:
+   :special-members:
+
+Jacobian Saliency Map Attack (JSMA)
+-----------------------------------
+.. autoclass:: SaliencyMapMethod
+   :members:
+   :special-members:
+
+NewtonFool
+----------
+.. autoclass:: NewtonFool
+   :members:
+   :special-members:
+
+Spatial Transformations Attack
+------------------------------
+.. autoclass:: SpatialTransformation
+   :members:
+   :special-members:
+
+Universal Perturbation Attack
+-----------------------------
+.. autoclass:: UniversalPerturbation
+   :members:
+   :special-members:
+
+Virtual Adversarial Method
+--------------------------
+.. autoclass:: VirtualAdversarialMethod
+   :members:
+   :special-members:
+
+Zeroth-Order Optimization Attack (ZOO)
+--------------------------------------
+.. autoclass:: ZooAttack
+   :members:
+   :special-members:
+
+HopSkipJump Attack
+------------------
+.. autoclass:: HopSkipJump
+   :members:
+   :special-members:
+
+Decision Tree Attack
+--------------------
+.. autoclass:: DecisionTreeAttack
+   :members:
+   :special-members:
+
+High Confidence Low Uncertainty Attack
+--------------------------------------
+.. autoclass:: HighConfidenceLowUncertainty
+   :members:
+   :special-members:

--- a/docs/modules/attacks/extraction.rst
+++ b/docs/modules/attacks/extraction.rst
@@ -1,0 +1,15 @@
+:mod:`art.attacks.extraction`
+=============================
+.. automodule:: art.attacks.extraction
+
+CopyCat CNN
+-----------
+.. autoclass:: CopycatCNN
+   :members:
+   :special-members:
+
+Functionally Equivalent Extraction
+----------------------------------
+.. autoclass:: FunctionallyEquivalentExtraction
+   :members:
+   :special-members:

--- a/docs/modules/attacks/poisoning.rst
+++ b/docs/modules/attacks/poisoning.rst
@@ -1,0 +1,9 @@
+:mod:`art.attacks.poisoning`
+============================
+.. automodule:: art.attacks.poisoning
+
+Poisoning SVM Attack
+--------------------
+.. autoclass:: PoisoningAttackSVM
+   :members:
+   :special-members:

